### ISSUE TITLE
angular11-microfrontends-lazy-components production build not working

### ISSUE THREAD

**Issue by cristobaldominguez95** [module-federation/module-federation-examples#938]
Hi,

I'm facing an error when I try to load a remote component in the production build. For some reason, it works in development mode when I serve the application, but after I generate the build and try to run it from a local http server I get the following error in the console:

```
main.99097cdd2ad1485c22e3.js:1 ERROR Error: Uncaught (in promise): TypeError: Cannot read property 'ɵcmp' of undefined
TypeError: Cannot read property 'ɵcmp' of undefined
```

I have debugged and realised that the Angular NgModule that I'm remotely loading in the production build, has the 'exports' property set as an empty array. However in the code it's exporting the component, and you can see it's there in the development version. I don't know if there's something I should change in the webpack configuration or something, I've been trying out things for a few days and there's no way I could get it to work...

Any help would be really appreciated.

Thank you

**Comment by ScriptedAlchemy** [comment]
Sorry I have no angular experience and don’t know how to make angular 11 work. Check the git blame and see if the user who PRd it can help, tag them in this issue

**Comment by shashikiran-sk** [comment]
I'm also facing the same error. Someone please help if there is some solution already

**Comment by michielkikkert** [comment]
Maybe this helps someone? 

https://github.com/angular/angular/issues/36279

**Comment by nohanna** [comment]
I had the same issue, I fixed it by adding a static field in my module to flag my components as "used" as suggested by the Angular team [here](https://github.com/angular/angular-cli/issues/21952#issuecomment-945752429).

Don't forget that in production, your components will have an uglyfied name so I found a workaround to add the name of my component by adding the [exportAs](https://angular.io/api/core/Directive#exportAs) field in my components annotations.
I also created a working example [here](https://github.com/nohanna/angular-micro-frontends).

Hope it helps!

**Comment by noxborough** [comment]
@nohanna 
Using Angular 11 (with AOT enabled)
I have a very similar issue for a dynamic component module bundle I'm generating. When I set **optimization: true** in my angular.json. The module gets built, but the declarations are magically omitted. Works just fine when **optimization: false**.

The **ɵmod.declarations** property inside the module I load has an empty array. And the component code is completely removed from the bundle. 

I followed the recommendation to add a static declarations variable in the module class. That did indeed end up preserving the component code inside the bundle. But **emod.declarations** is still empty.  Did you have the same experience?

**Comment by nohanna** [comment]
@noxborough If you are using AOT compilation, adding a static field will not change anything about `emod.declarations`, it will be still empty. It will simply add a new field (that will not be tree-shaked) to your module. So you have to retrieve your components from your static field not from `emod.declarations`.

**Comment by noxborough** [comment]
@nohanna 
I saw your other recommendation to use **exportAs** as a way to reference your components by a "known" name after optimization.  This does work, but for my use case I need to maintain the class names even after uglify/minify process. 

So in my custom webpack.config.js I include this section (see below). The **keep_classnames** is the key there.  But after I included this, I also had to set **concatenateModules: false,** because for some reason the class names ended up being 
doubled up (i.e. if class name was **ClassAComponent**, it would end up as **ClassAComponentClassAComponent**). Very weird, but found the solution here: https://stackoverflow.com/questions/49999999/duplicate-class-names-after-using-webpack-in-production-mode-with-uglifys-keep

```
optimization: {
        concatenateModules: false,
        minimizer: [
            (compiler) => {
                compiler.options.optimization.minimizer.filter(({ constructor: { name } }) => name === "TerserPlugin").forEach(
                    terser => {
                        terser.options.terserOptions.keep_classnames = true;
                        terser.options.terserOptions.keep_fnames = false;
                    }
                );
            },
        ],
    },
```
